### PR TITLE
Enable PublicApiAnalyzers

### PR DESCRIPTION
**Pull request by keegan-caruso**
# Enable PublicApiAnalyzers

See: https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md

Include internals for assemblies that are marked with InternalsVisibleTo

### CODE REVIEW

**Comment by keegan-caruso** [comment]
AzDO is passing but not the github action, interesting.

**Comment by jennyf19** [comment]
> AzDO is passing but not the github action, interesting.

You might try updating the github action from preview 4 to 7 for .NET 9

9.0.100-preview.7.24407.12

**Comment by pmaytak** [comment]
> As we added this mid release, I had to manually adjust shipped vs unshipped using [this as reference](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/compare/8.0.1...dev)

What something like this work - checkout the last 8.0.1 commit, run the tool to generate text files, stash them, and move them over to this branch?

How and when are those files generated, especially the for the unshipped API? Does the tool run automatically somehow or manually update the files?

File: .editorconfig [hunk 1]
```diff
@@ -157,6 +157,14 @@ csharp_space_between_square_brackets = false
 dotnet_code_quality.ca1802.api_surface = private, internal
 dotnet_code_quality.CA2208.api_surface = public
 
+dotnet_diagnostic.RS0051.severity = error
```
Comment (pmaytak) [inline .editorconfig:161]:
```suggestion
# https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/Microsoft.CodeAnalysis.PublicApiAnalyzers.md
dotnet_diagnostic.RS0051.severity = error
```

File: .editorconfig [hunk 2]
```diff
@@ -157,6 +157,14 @@ csharp_space_between_square_brackets = false
...
+dotnet_diagnostic.RS0052.severity = error
+dotnet_diagnostic.RS0053.severity = error
+dotnet_diagnostic.RS0054.severity = error
+dotnet_diagnostic.RS0055.severity = error
+dotnet_diagnostic.RS0057.severity = error
+dotnet_diagnostic.RS0058.severity = error
```
Comment (pmaytak) [inline .editorconfig:167]: Rules 22 and 61 seem useful.

File: src/Microsoft.IdentityModel.JsonWebTokens/Microsoft.IdentityModel.JsonWebTokens.csproj [hunk 3]
```diff
@@ -32,4 +32,14 @@
     <Reference Include="Microsoft.CSharp" />
   </ItemGroup>
 
+  <ItemGroup>
+    <AdditionalFiles Include="InternalAPI.Shipped.txt" />
+    <AdditionalFiles Include="InternalAPI.Unshipped.txt" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <AdditionalFiles Include="PublicAPI/$(TargetFramework)/InternalAPI.Shipped.txt" />
```
Comment (pmaytak) [inline src/Microsoft.IdentityModel.JsonWebTokens/Microsoft.IdentityModel.JsonWebTokens.csproj:45]: The approach this analyzer takes makes the folder structure a bit complex :(

File: build/common.props [hunk 4]
```diff
@@ -73,6 +73,13 @@
     </PackageReference>
   </ItemGroup>
 
+  <ItemGroup>
+    <PackageReference Include="Microsoft.CodeAnalysis.PublicApiAnalyzers" Version="3.3.4">
```
Comment (pmaytak) [inline build/common.props:72]: Do the latest preview versions have any relevant improvements? Didn't see a changelog for them.

https://www.nuget.org/packages/Microsoft.CodeAnalysis.PublicApiAnalyzers/3.11.0-beta1.24324.1

Comment (keegan-caruso) [reply build/common.props:72]: Same, couldn't find a reason to take the preview versions.

File: src/Microsoft.IdentityModel.JsonWebTokens/Microsoft.IdentityModel.JsonWebTokens.csproj (hunk 3 above)
Comment (keegan-caruso) [reply src/Microsoft.IdentityModel.JsonWebTokens/Microsoft.IdentityModel.JsonWebTokens.csproj:45]: Agreed

File: .editorconfig (hunk 2 above)
Comment (keegan-caruso) [reply .editorconfig:167]: 22 is enabled by default, I will add 61

**Comment by keegan-caruso** [comment]
> What something like this work - checkout the last 8.0.1 commit, run the tool to generate text files, stash them, and move them over to this branch?
> 
> How and when are those files generated, especially the for the unshipped API? Does the tool run automatically somehow or manually update the files?

These are generated by the fixer with the analyzer. But it doesn't work great for per target items. You can generate it, but then have to manually move it. 

I think it should be okay as it. At the next release, we move everything over to released.

**Review by jennyf19 — APPROVED** [review]
:shipit:

**Review by pmaytak — APPROVED** [review]

**Comment by keegan-caruso** [comment]
> > As we added this mid release, I had to manually adjust shipped vs unshipped using [this as reference](https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/compare/8.0.1...dev)
> 
> What something like this work - checkout the last 8.0.1 commit, run the tool to generate text files, stash them, and move them over to this branch?
> 
> How and when are those files generated, especially the for the unshipped API? Does the tool run automatically somehow or manually update the files?

I ended up going a different route, just start with everything unshipped, on the first release using this, move everything to shipped

**Comment by jennyf19** [comment]
@keegan-caruso do you want this in?

**Review by kllysng — APPROVED** [review]

**Review by kllysng — APPROVED** [review]

**Pull request merged**